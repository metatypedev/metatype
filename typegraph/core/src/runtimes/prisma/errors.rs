// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::wit::core::Error;

// pub fn relationship_not_found(source_model: &str, field: &str) -> Error {
//     format!("relationship target not found for  {source_model}::{field}")
// }

pub fn ambiguous_side(
    first_model: &str,
    first_field: &str,
    second_model: &str,
    second_field: &str,
) -> Error {
    [
        "Ambiguous side:",
        "cannot determine which model should have the foreign key:",
        &format!(
            "{}::{} or {}::{}",
            first_model, first_field, second_model, second_field
        ),
        "Please add 'fkey' or 'unique' attribute on one side",
    ]
    .join(" ")
    .into()
}

pub fn conflicting_attributes(
    attribute: &str,
    first_model: &str,
    first_field: &str,
    second_model: &str,
    second_field: &str,
) -> Error {
    [
        format!("Conflicting attributes '{attribute}':"),
        format!("on {first_model}::{first_field}"),
        format!("and {second_model}::{second_field}"),
    ]
    .join(" ")
    .into()
}

pub fn no_relationship_target(model: &str, field: &str, target_model: &str) -> Error {
    format!(r#"Relationship target field not found for "{model}::{field}" on {target_model:?}."#)
        .into()
}
