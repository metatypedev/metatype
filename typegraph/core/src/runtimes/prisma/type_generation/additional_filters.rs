// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::runtimes::prisma::context::PrismaContext;
use crate::runtimes::prisma::model::Property;
use crate::t::{self, ConcreteTypeBuilder, TypeBuilder};
use crate::{errors::Result, types::TypeId};

use super::TypeGen;

pub struct Take;

impl TypeGen for Take {
    fn generate(&self, _context: &PrismaContext) -> Result<TypeId> {
        t::integer().x_min(0).named(self.name()).build()
    }

    fn name(&self) -> String {
        "_Take".to_string()
    }
}

pub struct Skip;

impl TypeGen for Skip {
    fn generate(&self, _context: &PrismaContext) -> Result<TypeId> {
        t::integer().x_min(0).named(self.name()).build()
    }

    fn name(&self) -> String {
        "_Skip".to_string()
    }
}

pub struct Distinct(pub TypeId);

impl TypeGen for Distinct {
    fn generate(&self, context: &PrismaContext) -> Result<TypeId> {
        let model = context.model(self.0)?;
        let model = model.borrow();

        let cols: Vec<_> = model
            .iter_props()
            .filter_map(|(k, prop)| match prop {
                Property::Scalar(_) | Property::Model(_) => Some(k.to_string()),
                Property::Unmanaged(_) => None,
            })
            .collect();

        t::listx(t::string().enum_(cols))?
            .named(self.name())
            .build()
    }

    fn name(&self) -> String {
        let model_name = self.0.type_name().unwrap().unwrap();
        format!("_KeysOf_{model_name}")
    }
}
