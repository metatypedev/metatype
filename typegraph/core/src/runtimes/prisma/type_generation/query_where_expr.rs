// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::global_store::with_store;
use crate::t::{self, ConcreteTypeBuilder, TypeBuilder};
use crate::{errors::Result, types::TypeId};

use super::{where_::Where, with_filters::WithFilters, TypeGen};

pub struct QueryWhereExpr {
    model_id: TypeId,
}

impl QueryWhereExpr {
    pub fn new(model_id: TypeId) -> Self {
        Self { model_id }
    }
}

impl TypeGen for QueryWhereExpr {
    fn generate(&self, context: &mut super::TypeGenContext) -> Result<TypeId> {
        let where_type = context.generate(&Where::new(self.model_id, true))?;
        let extended_type =
            context.generate(&WithFilters::new(where_type, self.model_id, false))?;

        let props = with_store(|s| extended_type.as_struct(s).unwrap().data.props.to_vec());

        let name = self.name(context);
        let self_ref = t::proxy(&name).build()?;
        let and = t::optional(t::array(self_ref).build()?).build()?;

        let mut builder = t::struct_();
        builder.named(name);
        for (k, ty) in props.into_iter() {
            builder.prop(k.clone(), ty.into());
        }
        builder.prop("AND", and);
        builder.prop("OR", and);
        builder.prop("NOT", t::optional(self_ref).build()?);

        builder.build()
    }

    fn name(&self, context: &super::TypeGenContext) -> String {
        format!(
            "Query{}WhereInput",
            context.registry.models.get(&self.model_id).unwrap().name
        )
    }
}
