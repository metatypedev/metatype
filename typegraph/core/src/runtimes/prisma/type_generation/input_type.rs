// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::errors::Result;
use crate::runtimes::prisma::type_utils::RuntimeConfig;
use crate::runtimes::prisma::{relationship::Cardinality, type_generation::where_::Where};
use crate::t::{self, ConcreteTypeBuilder, TypeBuilder};
use crate::types::{Type, TypeId};

use super::{TypeGen, TypeGenContext};

#[derive(Clone, Copy)]
enum Operation {
    Create,
    Update,
}

impl Operation {
    fn is_update(&self) -> bool {
        match self {
            Operation::Create => false,
            Operation::Update => true,
        }
    }
}

pub struct InputType {
    model_id: TypeId,
    skip_rel: Vec<String>,
    operation: Operation,
}

impl InputType {
    pub fn for_create(model_id: TypeId) -> Self {
        Self {
            model_id,
            skip_rel: vec![],
            operation: Operation::Create,
        }
    }

    pub fn for_update(model_id: TypeId) -> Self {
        Self {
            model_id,
            skip_rel: vec![],
            operation: Operation::Update,
        }
    }
}

impl TypeGen for InputType {
    fn generate(&self, context: &mut TypeGenContext) -> Result<TypeId> {
        let mut builder = t::struct_();

        for (k, type_id) in self.model_id.as_struct()?.iter_props() {
            let rel = context.registry.find_relationship_on(self.model_id, k);
            if let Some(rel) = rel {
                if self.skip_rel.contains(&rel.name) {
                    continue;
                }

                let entry = rel.get(rel.side_of_type(type_id).unwrap());
                // model
                let create = context.generate(&InputType {
                    model_id: entry.model_type,
                    skip_rel: {
                        let mut skip_rel = self.skip_rel.clone();
                        skip_rel.push(rel.name.clone());
                        skip_rel
                    },
                    operation: Operation::Create,
                })?;
                // TODO unique where
                let connect = context.generate(&Where::new(entry.model_type, false))?;

                let mut inner = t::struct_();
                inner.propx("create", t::optional(create))?;
                inner.propx("connect", t::optional(connect))?;

                if let Cardinality::Many = entry.cardinality {
                    inner.propx(
                        "createMany",
                        t::optionalx(t::struct_().propx("data", t::array(create))?)?,
                    )?;
                }

                // TODO what if cardinality is Cardinality::One ??
                builder.propx(k, t::optionalx(inner.min(1).max(1))?)?;
            } else {
                let attrs = type_id.attrs()?;
                match attrs.concrete_type.as_type()? {
                    Type::Func(_) => continue,
                    typ => {
                        let is_auto = || -> Result<_> {
                            Ok(RuntimeConfig::try_from(&typ)?.get("auto")?.unwrap_or(false))
                        };
                        builder.prop(
                            k,
                            if self.operation.is_update() || is_auto()? {
                                t::optional(type_id).build()?
                            } else {
                                type_id
                            },
                        );
                    }
                }
            }
        }

        builder.named(self.name()).build()
    }

    fn name(&self) -> String {
        let model_name = self.model_id.type_name().unwrap().unwrap();
        let suffix = if self.skip_rel.is_empty() {
            "".to_string()
        } else {
            format!("_excluding_{}", self.skip_rel.join("And"))
        };
        let op = match self.operation {
            Operation::Create => "Create",
            Operation::Update => "Update",
        };
        format!("_{model_name}_{op}Input{suffix}")
    }
}
