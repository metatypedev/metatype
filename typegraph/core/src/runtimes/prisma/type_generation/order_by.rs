// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::errors::Result;
use crate::runtimes::prisma::relationship::Cardinality;
use crate::t::{ConcreteTypeBuilder, TypeBuilder};
use crate::types::Type;
use crate::{t, types::TypeId};

use super::{TypeGen, TypeGenContext};

pub struct OrderBy {
    model_id: TypeId,
    skip_rel: Vec<String>,
    aggregates: bool,
}

impl OrderBy {
    pub fn new(model_id: TypeId) -> Self {
        Self {
            model_id,
            skip_rel: vec![],
            aggregates: false,
        }
    }

    pub fn skip(mut self, rel_names: Vec<String>) -> Self {
        self.skip_rel = rel_names;
        self
    }

    pub fn with_aggregates(self) -> Self {
        Self {
            aggregates: true,
            ..self
        }
    }
}

impl TypeGen for OrderBy {
    fn generate(&self, context: &mut TypeGenContext) -> Result<TypeId> {
        let mut builder = if self.aggregates {
            t::struct_extends(context.generate(&AggregateSorting::new(self.model_id))?)?
        } else {
            t::struct_()
        };

        for (k, type_id) in self.model_id.as_struct()?.iter_props() {
            let registry_entry = context.registry.models.get(&self.model_id).unwrap();
            let rel = registry_entry.relationships.get(k);
            if let Some(rel_name) = rel {
                if self.skip_rel.contains(rel_name) {
                    continue;
                }

                let rel = context
                    .registry
                    .relationships
                    .get(rel_name)
                    .cloned()
                    .unwrap();

                // TODO does this work for self relationship?
                let relationship_model = rel.get_opposite_of(self.model_id, k).unwrap();
                match relationship_model.cardinality {
                    Cardinality::Many => {
                        builder.prop(k, context.generate(&SortByAggregates)?);
                    }
                    Cardinality::Optional | Cardinality::One => {
                        let mut skip_rel = self.skip_rel.clone();
                        skip_rel.push(rel_name.clone());
                        let inner = context.generate(
                            &OrderBy::new(relationship_model.model_type).skip(skip_rel),
                        )?;
                        builder.prop(k, t::optional(inner).build()?);
                    }
                }
            } else {
                let typ = type_id.as_type()?;
                let (typ, nullable) = if let Type::Optional(inner) = typ {
                    (inner.item().as_type()?, true)
                } else {
                    (typ, false)
                };
                match typ {
                    Type::Boolean(_) | Type::Integer(_) | Type::Float(_) | Type::String(_) => {
                        builder.prop(k, context.generate(&Sort { nullable })?);
                    }
                    Type::Array(_) => {}
                    _ => Err(format!("Cannot order by type {:?}", typ))?,
                }
            }
        }

        t::array(builder.build()?).named(self.name()).build()
    }

    fn name(&self) -> String {
        let name = self.model_id.type_name().unwrap().unwrap();
        let suffix = if self.skip_rel.is_empty() {
            "".to_string()
        } else {
            format!("_excluding_{}", self.skip_rel.join("_"))
        };
        let suffix2 = if self.aggregates {
            "_with_aggregates"
        } else {
            ""
        };
        format!("_{}_OrderBy{suffix}{suffix2}", name)
    }
}

struct SortOrder;

impl TypeGen for SortOrder {
    fn generate(&self, _context: &mut TypeGenContext) -> Result<TypeId> {
        t::string()
            .enum_(vec!["asc".to_string(), "desc".to_string()])
            .named(self.name())
            .build()
    }

    fn name(&self) -> String {
        "_SortOrder".to_string()
    }
}

struct NullsOrder;

impl TypeGen for NullsOrder {
    fn generate(&self, _context: &mut TypeGenContext) -> Result<TypeId> {
        t::string()
            .enum_(vec!["first".to_string(), "last".to_string()])
            .named(self.name())
            .build()
    }

    fn name(&self) -> String {
        "_NullsOrder".to_string()
    }
}

struct Sort {
    nullable: bool,
}

impl TypeGen for Sort {
    fn generate(&self, context: &mut TypeGenContext) -> Result<TypeId> {
        let sort_order = context.generate(&SortOrder)?;
        let nulls_order = context.generate(&NullsOrder)?;
        let mut builder = t::struct_();
        builder.prop("sort", sort_order);
        if self.nullable {
            builder.prop("nulls", nulls_order);
        }

        t::optional(t::union([builder.build()?, sort_order]).build()?)
            .named(self.name())
            .build()
    }

    fn name(&self) -> String {
        let nullable = if self.nullable { "_nullable" } else { "" };
        format!("_Sort{}", nullable)
    }
}

struct SortByAggregates;

impl TypeGen for SortByAggregates {
    fn generate(&self, context: &mut TypeGenContext) -> Result<TypeId> {
        let mut builder = t::struct_();
        let sort = context.generate(&Sort { nullable: false })?;
        builder.prop("_count", sort);
        builder.prop("_avg", sort);
        builder.prop("_sum", sort);
        builder.prop("_min", sort);
        builder.prop("_max", sort);

        t::optional(builder.build()?).named(self.name()).build()
    }

    fn name(&self) -> String {
        "_SortByAggregates".to_string()
    }
}

struct AggregateSorting {
    model_id: TypeId,
}

impl AggregateSorting {
    fn new(model_id: TypeId) -> Self {
        Self { model_id }
    }
}

impl TypeGen for AggregateSorting {
    fn generate(&self, context: &mut TypeGenContext) -> Result<TypeId> {
        use AggregateSortingProp as Prop;

        let props = self
            .model_id
            .as_struct()?
            .iter_props()
            .map(|(k, type_id)| -> Result<_> {
                let attrs = type_id.attrs()?;
                let typ = attrs.concrete_type.as_type()?;
                let (typ, is_optional) = match typ {
                    Type::Optional(inner) => (
                        TypeId(inner.data.of).attrs()?.concrete_type.as_type()?,
                        true,
                    ),
                    _ => (typ, false),
                };
                match typ {
                    Type::Integer(_) | Type::Float(_) => Ok(Prop {
                        key: k.to_string(),
                        number_type: true,
                        optional: is_optional,
                    }),
                    _ => Ok(Prop {
                        key: k.to_string(),
                        number_type: false,
                        optional: is_optional,
                    }),
                }
            })
            .collect::<Result<Vec<_>>>()?;

        let mut builder = t::struct_();
        let count = t::optional(
            t::struct_from(props.iter().filter_map(|p| {
                p.generate(context, true)
                    .unwrap()
                    .map(|ty| (p.key.clone(), ty))
            }))
            .build()?,
        )
        .build()?;
        let others = t::optional(
            t::struct_from(props.iter().filter_map(|p| {
                p.generate(context, false)
                    .unwrap()
                    .map(|ty| (p.key.clone(), ty))
            }))
            .build()?,
        )
        .build()?;
        builder
            .prop("_count", count)
            .prop("_avg", others)
            .prop("_sum", others)
            .prop("_min", others)
            .prop("_max", others)
            .named(self.name())
            .build()
    }

    fn name(&self) -> String {
        let model_name = self.model_id.type_name().unwrap().unwrap();
        format!("_{model_name}_AggregateSorting")
    }
}

struct AggregateSortingProp {
    key: String,
    number_type: bool,
    optional: bool,
}

impl AggregateSortingProp {
    fn generate(&self, context: &mut TypeGenContext, count: bool) -> Result<Option<TypeId>> {
        Ok(match count {
            true => Some(context.generate(&Sort {
                nullable: self.optional,
            })?),
            false => {
                if self.number_type {
                    Some(context.generate(&Sort {
                        nullable: self.optional,
                    })?)
                } else {
                    None
                }
            }
        })
    }
}
