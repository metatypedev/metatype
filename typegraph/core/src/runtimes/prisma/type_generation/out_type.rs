// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::errors::Result;
use crate::global_store::with_store;
use crate::runtimes::prisma::type_utils::get_type_name;
use crate::t::{self, ConcreteTypeBuilder, TypeBuilder};
use crate::types::{Type, TypeId};

use super::{Cardinality, TypeGen};

pub struct OutType {
    model_id: TypeId,
    skip_rel: Vec<String>, // list of relationships to skip to avoid infinite recursion
}

impl OutType {
    pub fn new(model_id: TypeId) -> Self {
        Self {
            model_id,
            skip_rel: vec![],
        }
    }
}

impl TypeGen for OutType {
    fn generate(
        &self,
        context: &mut super::TypeGenContext,
    ) -> crate::errors::Result<crate::types::TypeId> {
        enum PropType {
            Scalar(TypeId),
            Model {
                model_id: TypeId,
                cardinality: Cardinality,
                rel_name: String,
            },
        }
        struct Prop {
            key: String,
            typ: PropType,
        }
        let mut props = vec![];

        with_store(|s| -> Result<_> {
            for (k, id) in self.model_id.as_struct(s).unwrap().data.props.iter() {
                let rel = context.registry.find_relationship_on(self.model_id, k);
                if let Some(rel) = rel {
                    if self.skip_rel.contains(&rel.name)
                        || rel.left.model_type == rel.right.model_type
                    {
                        continue;
                    }
                    let entry = rel.get(rel.side_of_type(id.into()).unwrap());
                    props.push(Prop {
                        key: k.to_string(),
                        typ: PropType::Model {
                            model_id: entry.model_type,
                            cardinality: entry.cardinality,
                            rel_name: rel.name.clone(),
                        },
                    })
                } else {
                    match s.get_attributes(id.into())?.concrete_type.as_type(s)? {
                        Type::Func(_) => {
                            // skip, other runtime
                            continue;
                        }
                        _ => props.push(Prop {
                            key: k.to_string(),
                            typ: PropType::Scalar(id.into()),
                        }),
                    }
                }
            }
            Ok(())
        })?;

        let mut builder = t::struct_();
        builder.named(self.name(context));

        for prop in props.into_iter() {
            match prop.typ {
                PropType::Scalar(id) => {
                    builder.prop(prop.key, id);
                }
                PropType::Model {
                    model_id,
                    cardinality,
                    rel_name,
                } => {
                    let mut skip_rel = self.skip_rel.clone();
                    skip_rel.push(rel_name);

                    let ty = context.generate(&OutType { model_id, skip_rel })?;
                    let ty = match cardinality {
                        Cardinality::Optional => t::optional(ty).build()?,
                        Cardinality::One => ty,
                        Cardinality::Many => t::array(ty).build()?,
                    };

                    builder.prop(prop.key, ty);
                }
            }
        }

        builder.build()
    }

    fn name(&self, _context: &super::TypeGenContext) -> String {
        let model_name = get_type_name(self.model_id).expect("prisma model must have a name");
        let suffix = if self.skip_rel.is_empty() {
            String::new()
        } else {
            format!("_excluding_{}", self.skip_rel.join("_"))
        };
        format!("_{model_name}OutputType{suffix}")
    }
}
