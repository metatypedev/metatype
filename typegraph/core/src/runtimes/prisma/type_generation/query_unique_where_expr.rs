// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::errors::Result;
use crate::t::{self, ConcreteTypeBuilder, TypeBuilder};
use crate::types::TypeFun;
use crate::{global_store::with_store, types::TypeId};

use super::{TypeGen, TypeGenContext};

pub struct QueryUniqueWhereExpr {
    model_id: TypeId,
}

impl QueryUniqueWhereExpr {
    pub fn new(model_id: TypeId) -> Self {
        Self { model_id }
    }
}

impl TypeGen for QueryUniqueWhereExpr {
    fn generate(&self, context: &mut TypeGenContext) -> Result<TypeId> {
        let props = with_store(|s| -> Result<_> {
            let model = s.type_as_struct(self.model_id).unwrap();

            let mut props = vec![];
            for (k, ty) in model.data.props.iter() {
                let attrs = s.get_attributes(ty.into())?;
                let is_id = attrs
                    .concrete_type
                    .as_type(s)?
                    .get_base()
                    .ok_or_else(|| "expected a concrete type".to_string())?
                    .as_id;
                let is_unique = s.get_type(ty.into())?.get_base().map_or(false, |base| {
                    base.runtime_config
                        .iter()
                        .flatten()
                        .find_map(|(k, v)| (k == "unique").then(|| v.clone()))
                        .map_or(false, |v| v == "true")
                });
                if s.is_func(attrs.concrete_type)? || (!is_id && !is_unique) {
                    continue;
                }
                props.push((k.clone(), s.resolve_quant(attrs.concrete_type)?));
            }

            Ok(props)
        })?;

        let mut st = t::struct_();
        for (k, ty) in props.into_iter() {
            st.prop(k, t::optional(ty).build()?);
        }
        st.named(self.name(context)).build()
    }

    fn name(&self, _context: &TypeGenContext) -> String {
        let name = with_store(|s| s.get_type_name(self.model_id).unwrap().unwrap().to_string());
        format!("QueryUnique{}WhereInput", name)
    }
}
