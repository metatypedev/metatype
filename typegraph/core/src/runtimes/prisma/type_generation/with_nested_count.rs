// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::errors::Result;
use crate::runtimes::prisma::type_generation::count::Count;
use crate::t::{self, ConcreteTypeBuilder, TypeBuilder};
use crate::types::ProxyResolution;
use crate::{runtimes::prisma::relationship::Cardinality, types::TypeId};

use super::{TypeGen, TypeGenContext};

pub struct WithNestedCount {
    model_id: TypeId,
    skip: Vec<String>,
}

impl WithNestedCount {
    pub fn new(model_id: TypeId) -> Self {
        Self {
            model_id,
            skip: vec![],
        }
    }
}

impl TypeGen for WithNestedCount {
    fn generate(&self, context: &mut TypeGenContext) -> Result<TypeId> {
        let mut countable = vec![];
        let mut builder = t::struct_();

        let model = self.model_id.as_struct().unwrap();
        for (key, type_id) in model.iter_props() {
            if let Some(rel) = context.registry.find_relationship_on(self.model_id, key) {
                if self.skip.contains(&rel.name) {
                    continue;
                }
                // TODO get_concrete_type(resolve_proxy=true)
                let relation_model = rel.get_opposite_of(self.model_id, key).unwrap();
                match relation_model.cardinality {
                    Cardinality::Optional => {
                        let inner = context.generate(&WithNestedCount {
                            model_id: relation_model.model_type,
                            skip: [self.skip.as_slice(), &[rel.name.clone()]].concat(),
                        })?;
                        builder.propx(key, t::optional(inner))?;
                        countable.push(key.to_string());
                    }
                    Cardinality::Many => {
                        let inner = context.generate(&WithNestedCount {
                            model_id: relation_model.model_type,
                            skip: [self.skip.as_slice(), &[rel.name.clone()]].concat(),
                        })?;
                        builder.propx(key, t::array(inner))?;
                        countable.push(key.to_string());
                    }
                    Cardinality::One => {
                        builder.prop(key, relation_model.model_type);
                    }
                }
            } else {
                let type_id = type_id.concrete_type(ProxyResolution::Force)?.unwrap();
                builder.prop(key, type_id);
            }
        }

        if !countable.is_empty() {
            let mut count = t::struct_();
            for prop in countable.into_iter() {
                count.prop(prop, context.generate(&Count)?);
            }
            builder.propx("_count", count)?;
        }

        builder.named(self.name()).build()
    }

    fn name(&self) -> String {
        let model_name = self.model_id.type_name().unwrap().unwrap();
        let suffix = if self.skip.is_empty() {
            "".to_string()
        } else {
            format!("_excluding_{}", self.skip.join("_"))
        };
        format!("{model_name}WithNestedCount{suffix}")
    }
}
