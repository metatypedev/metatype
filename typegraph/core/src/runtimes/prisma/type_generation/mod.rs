// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use std::collections::hash_map::Entry;
use std::collections::HashMap;

use self::additional_filters::{Distinct, Skip, Take};
use self::input_type::InputType;
use self::order_by::OrderBy;
use self::out_type::OutType;
use self::query_unique_where_expr::QueryUniqueWhereExpr;
use self::query_where_expr::QueryWhereExpr;
use self::with_nested_count::WithNestedCount;

use super::relationship::registry::RelationshipRegistry;
use crate::errors::Result;
use crate::global_store::with_store;
use crate::runtimes::prisma::relationship::Cardinality;
use crate::t::{self, TypeBuilder};
use crate::types::{ProxyResolution, Type, TypeFun, TypeId};

mod additional_filters;
mod count;
mod input_type;
mod order_by;
mod out_type;
mod query_unique_where_expr;
mod query_where_expr;
mod where_;
mod with_filters;
mod with_nested_count;

#[derive(Default, Debug)]
pub struct TypeGenContext {
    pub registry: RelationshipRegistry,
    cache: HashMap<String, TypeId>,
}

trait TypeGen {
    fn generate(&self, context: &mut TypeGenContext) -> Result<TypeId>;
    fn name(&self, context: &TypeGenContext) -> String;
}

impl TypeGenContext {
    pub fn find_unique(&mut self, model_id: TypeId) -> Result<(TypeId, TypeId)> {
        self.registry.manage(model_id)?;

        Ok((
            // input
            t::struct_()
                .prop(
                    "where",
                    t::optional(self.generate(&QueryUniqueWhereExpr::new(model_id))?).build()?,
                )
                .build()?,
            // output
            t::optional(self.generate(&WithNestedCount::new(model_id))?).build()?,
        ))
    }

    fn find_many_inp(&mut self, model_id: TypeId) -> Result<TypeId> {
        Ok(t::struct_()
            .prop(
                "where",
                t::optional(self.generate(&QueryWhereExpr::new(model_id))?).build()?,
            )
            .prop(
                "orderBy",
                t::optional(self.generate(&OrderBy::new(model_id, vec![]))?).build()?,
            )
            .prop("take", t::optional(self.generate(&Take)?).build()?)
            .prop("skip", t::optional(self.generate(&Skip)?).build()?)
            .prop(
                "distinct",
                t::optional(self.generate(&Distinct(model_id))?).build()?,
            )
            .build()?)
    }

    pub fn find_many(&mut self, model_id: TypeId) -> Result<(TypeId, TypeId)> {
        self.registry.manage(model_id)?;

        Ok((
            // input
            self.find_many_inp(model_id)?,
            // output
            // t::integer().build()?
            t::array(self.generate(&WithNestedCount::new(model_id))?).build()?,
        ))
    }

    pub fn find_first(&mut self, model_id: TypeId) -> Result<(TypeId, TypeId)> {
        self.registry.manage(model_id)?;

        Ok((
            // input
            self.find_many_inp(model_id)?,
            // output
            t::optional(self.generate(&OutType::new(model_id))?).build()?,
        ))
    }

    pub fn create_one(&mut self, model_id: TypeId) -> Result<(TypeId, TypeId)> {
        self.registry.manage(model_id)?;

        Ok((
            // input
            t::struct_()
                .prop("data", self.generate(&InputType::new(model_id))?)
                .build()?,
            // output
            self.generate(&OutType::new(model_id))?,
        ))
    }

    fn generate(&mut self, generator: &impl TypeGen) -> Result<TypeId> {
        let type_name = generator.name(self);
        if let Some(type_id) = self.cache.get(&type_name) {
            Ok(*type_id)
        } else {
            let type_id = generator.generate(self)?;
            self.cache.insert(type_name, type_id);
            Ok(type_id)
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::test_utils::*;

    #[test]
    fn test_find_unique() -> Result<()> {
        let mut context = TypeGenContext::default();

        let (user, post) = models::simple_relationship()?;
        context.registry.manage(user)?;

        let (inp, out) = context.find_unique(user)?;
        insta::assert_snapshot!("find_unique User inp", tree::print(inp));
        insta::assert_snapshot!("find_unique User out", tree::print(out));

        let (inp, out) = context.find_unique(post)?;
        insta::assert_snapshot!("find_unique Post inp", tree::print(inp));
        insta::assert_snapshot!("find_unique Post out", tree::print(out));

        Ok(())
    }

    #[test]
    fn test_find_many() -> Result<()> {
        let mut context = TypeGenContext::default();

        let record = models::simple_record()?;
        context.registry.manage(record)?;
        let (inp, out) = context.find_many(record)?;
        insta::assert_snapshot!("find_many Record inp", tree::print(inp));
        insta::assert_snapshot!("find_many Record out", tree::print(out));

        let (user, post) = models::simple_relationship()?;
        context.registry.manage(user)?;

        let (inp, out) = context.find_many(user)?;
        insta::assert_snapshot!("find_many User inp", tree::print(inp));
        insta::assert_snapshot!("find_many User out", tree::print(out));

        let (inp, out) = context.find_many(post)?;
        insta::assert_snapshot!("find_many Post inp", tree::print(inp));
        insta::assert_snapshot!("find_many Post out", tree::print(out));

        Ok(())
    }

    #[test]
    fn test_find_first() -> Result<()> {
        // input types are the same as find_many's

        let mut context = TypeGenContext::default();

        let record = models::simple_record()?;
        context.registry.manage(record)?;
        let (_inp, out) = context.find_first(record)?;
        // insta::assert_snapshot!("find_first Record inp", tree::print(inp));
        insta::assert_snapshot!("find_first Record out", tree::print(out));

        let (user, post) = models::simple_relationship()?;
        context.registry.manage(user)?;

        let (_inp, out) = context.find_first(user)?;
        // insta::assert_snapshot!("find_first User inp", tree::print(inp));
        insta::assert_snapshot!("find_first User out", tree::print(out));

        let (_inp, out) = context.find_first(post)?;
        // insta::assert_snapshot!("find_first Post inp", tree::print(inp));
        insta::assert_snapshot!("find_first Post out", tree::print(out));

        Ok(())
    }

    #[test]
    fn test_create_one() -> Result<()> {
        let mut context = TypeGenContext::default();

        let record = models::simple_record()?;
        context.registry.manage(record)?;
        let (inp, out) = context.create_one(record)?;
        insta::assert_snapshot!("create_one Record inp", tree::print(inp));
        insta::assert_snapshot!("create_one Record out", tree::print(out));

        let (user, post) = models::simple_relationship()?;
        context.registry.manage(user)?;
        let (inp, out) = context.create_one(user)?;
        insta::assert_snapshot!("create_one User inp", tree::print(inp));
        insta::assert_snapshot!("create_one User out", tree::print(out));

        let (inp, out) = context.create_one(post)?;
        insta::assert_snapshot!("create_one Post inp", tree::print(inp));
        insta::assert_snapshot!("create_one Post out", tree::print(out));

        Ok(())
    }
}
