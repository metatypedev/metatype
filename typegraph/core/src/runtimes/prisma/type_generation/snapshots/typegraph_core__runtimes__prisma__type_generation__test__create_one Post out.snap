---
source: typegraph/core/src/runtimes/prisma/type_generation/mod.rs
expression: "tree::print(out)"
---
root: struct '_PostOutputType' #46
├─ [id]: integer #14
├─ [title]: string #15
└─ [author]: struct '_UserOutputType_excluding_UserToPost' #45
   ├─ [id]: integer #9
   └─ [name]: string #10

